README.md
# Ganger

[![Code Climate](https://codeclimate.com/github/andytinycat/ganger/badges/gpa.svg)](https://codeclimate.com/github/andytinycat/ganger)

## What is Ganger?

_ganger |ˈgaŋə| noun; Brit; the foreman of a gang of labourers._

Ganger is a tool for running short-lived network services inside Docker containers, and then
proxying client connections to them. Think of Ganger as a sort of 
multi-machine [(x)inetd](http://en.wikipedia.org/wiki/Xinetd). 

## How does Ganger work?

Ganger is very simple. It receives TCP connections from clients, creates a Docker container
with a network service (listening on TCP), then proxies the client connection to the Docker
container. When the connection is over, the Docker container is terminated.

Ganger can use one or more Docker daemons - so you can spread the containers over as many
machines as you wish. Ganger will use the least-utilised daemon (the one running the fewest 
containers) to launch a new machine.

It's not important what the network service is - only that you supply a Docker image which
was built with a port exposed via the `EXPOSE` command. You tell Ganger what that port is,
what the port you want to be available to clients on is, and Ganger will do the rest.

## That sounds mad! Why would you want to use Ganger?

The rationale for creating Ganger was driven by a real-world problem. At Forward3D we use
Hive to query the data we import into Hadoop. They run over small amounts of data.
[Hive local mode](http://hadoop-pig-hive-thejas.blogspot.co.uk/2013/04/running-hive-in-local-mode.html) is a
good way to run small jobs quickly without the overhead of launching MapReduce JVMs on a cluster.

However, local mode seems to leak memory in some situations, and drops tons of junk into /tmp
on the machine that eventually causes problems. Ganger was created to allow me to isolate
a single query to a single container, then throw it away when the query is complete. It
should be useful in similar situations where you want to run code in a totally pure environment
on every request.

Docker is the perfect tool for this, since the time and overhead required to start a Docker
container (when considering the time required to run a query) is minimal.

## Features

- Spreads the load over multiple Docker daemons, launching new containers on the least-utilised one.
- Tries to connect to the service port a configurable number of times (with configurable timeout);
  useful if you have a service that takes a while to start.
- Tells all configured Docker daemons to pull the image on startup.
- Control of various timeout and retry parameters.
- Discovery of Docker services with [Consul](http://consul.io). Now you can autoscale your
  docker hosts!
- Configurable max containers; will hold connections in a queue until they can be serviced.

## An example

You must have Docker installed (somewhere). The example config file assumes you're running
[Boot2Docker](https://github.com/boot2docker/boot2docker) on Mac OS X. If you're not, then you will 
need to modify the configuration file to add the URL to your Docker daemon. This can be local,
on some servers somewhere, in the cloud, etc.

Clone this repository, and `cd` into it, then run:

    bundle install
    bundle exec bin/ganger.rb

This will use the default config file, which will pull an example dummy 'service' that uses
ncat to echo back whatever you send it. This default image is hosted in
[my DockerHub repository](https://registry.hub.docker.com/u/andytinycat/ncat/), and just contains
ncat and will start it on 12345/tcp when run.

To test it, telnet to it (it defaults to listening on 5454):

    telnet localhost 5454

In another shell, run `docker ps`

    CONTAINER ID        IMAGE                     COMMAND                CREATED             STATUS              PORTS                      NAMES
    c8c438e47f96        andytinycat/ncat:latest   ncat -l 12345 -k -c    14 hours ago        Up 2 seconds        0.0.0.0:49176->12345/tcp   nostalgic_bohr

A Docker container was started to service the request.

When you're done, exit your telnet session (CTRL-], then type `quit` on Mac OS X), and run
`docker ps` again. The container will have shut down and been destroyed.

## Ganger inside a Docker container

There is a Docker image containing Ganger, if you prefer deploying containers. You can find
it on the [Docker Hub](https://registry.hub.docker.com/u/forward3d/ganger/).

It expects you to supply the path to the configuration file as an environment variable.
In order to allow the configuration file to live outside of the codebase inside the container,
you'll need to mount the file from the host filesystem into the container.

Here's an example of how to run the container:

    docker run \
      -v /etc/ganger/ganger-config.rb:/etc/ganger/ganger-config.rb \
      -e GANGER_CONFIG_FILE=/etc/ganger/ganger-config.rb \
      -p 10002:5454 \
      andytinycat/ganger

Note that if you're using Ganger in this way, you need to make sure your configuration file option
`ganger.listen_port` is set to the same port that you'll forward into the container. In the above case,
Ganger is listening on 5454/tcp inside the container, and forwarding 10002/tcp into it.

## Future features

Ganger is a simple proof-of-concept. However, the following features are planned:

 - Container reuse; keep a "container pool" and reuse containers a configurable number of times
 - A simple status webpage, similar to HAproxy
 - Graceful switchovers to a new container (for upgrades, etc)
 - Graceful cleanup of threads and containers when terminating
 - Detecting containers that exit before a request can be sent (indicates broken container)
 
## Contributing
 
Pull requests are welcomed! The current state of code works for the simple things I need it for,
but there's so much more that could be added (see features above, tests, etc).


CHANGELOG.md
# Ganger changelog

## 0.3.0

Significant redesign:

* A single thread is responsible for container allocation and launch, which makes
  features like pooling and limiting possible.
* The network code has been significantly hardened against various kinds of failures.
* The container allocation code has been separated off into 'engines'.
* Some basic barely-tested support for Consul discovery of Docker daemons has been added.
* The configuration file is now pure Ruby (like Vagrant etc).
* Numerous other fixes and improvements.

## 0.0.4

Protection against faulty container launches (see [Docker bug 7563](https://github.com/docker/docker/issues/7563)) that can occur under high concurrency.

## 0.0.2, 0.0.3

Various debugging changes.

## 0.0.1

First public release.


config/ganger-config.rb
Ganger.configure do |c|
  
  # The image - will be pulled on startup
  c.docker.image = "andytinycat/ncat"
  
  # The Docker command to run in the container; can be left blank, in which
  # case the image's command will be used
  c.docker.cmd = [ 'ncat', '-l', '12345', '-k', '-c', 'xargs -n1 echo']
  
  # The port to expose
  c.docker.expose = '12345/tcp'
  
  # The port we'll listen on for frontend connections
  c.ganger.listen_port = 5454
  
  # Try to connect immediately after starting the container, then
  # 6 more times, with a timeout of 5 seconds
  c.ganger.service_connection_timeout = 5
  c.ganger.service_connection_retries = 6
  
  # Wait 300 seconds for Docker to respond (image pulls can be slow!)
  c.ganger.docker_timeout = 300
  
  # Engine to use to locate Docker servers
  # Two engines are available: 'static' and 'consul'.
  # Static uses a list of pre-configured Docker daemons.
  # Consul uses the Consul service discovery tool (http://consul.io)
  # to locate Docker servers to use.
  c.ganger.docker_discovery = 'static'
  
  #
  # STATIC ENGINE SETTINGS
  #
  
  # These are all the Docker daemons we can start containers on
  # It must be an array of hashes, with each hash having a url key
  # and a max_containers key. If you want to allow an infinite number of
  # concurrent containers, specify -1. (This is not a good idea!)
  c.static_engine.daemons = [
    {
      url: 'boot2docker',
      max_containers: 2
    }
  ]
  
  #
  # CONSUL ENGINE SETTINGS
  #
  
  # These are the hosts running Consul that you want to use for
  # service discovery. You should have at least 3 of these Consul agents
  # running in server mode for availability.
  c.consul_engine.hosts = [ 'host1:8500', 'host2:8500', 'host3:8500' ]
  
  # The name of the service registered with Consul
  c.consul_engine.service_name = 'docker'
  
  # The maximum number of containers each discovered server will support,
  # if not specified in a tag (see the README)
  c.consul_engine.default_max_containers = 10
  
  # Consul has three consistency modes when making API request; see the Consul documentation
  # for more information (http://www.consul.io/docs/agent/http.html)
  # We default to 'stale', since this is fast and we can cope with service data being
  # slightly out of date.
  c.consul_engine.consistency_mode = 'stale'
  
  # List of datacenters to try for services, in order of preference.
  # e.g. Here we'd try to find a service in DC 'eu', then 'us', then 'faraway'
  c.consul_engine.datacenters = ['eu', 'us', 'faraway']
  
end